### PR TITLE
Update ci.yml to be dispatchable and skippable

### PR DESCRIPTION
**Pull request by jonathansadowski**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR updates `ci.yml` such that it is skipped (except via manual dispatch) when the repo variable `SKIP_CI` is set. This allows forks that want to enable the workflow for manual testing to do so without having it run on every PR.

The PR also adds a manual dispatch, which can either be triggered for the entire suite of tests in `ci.yml`, or, if a PR number is provided, it will be triggered as if it were run on that PR.

Closes # .

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a fork of this repo, and merge the changes in the PR to `trunk` within your fork.
2. Enable actions if they aren't already. You may want to disable other workflows that run on the PR so that you don't have to wait around for `ci.yml` to receive workers.
3. Create a new PR against `trunk` of your fork. Edit a file within one of the projects or plugins covered by `ci.yml`. For example, make an edit (such as adding a comment) to `packages/js/currency/src/index.ts`.
4. Verify that the project matrix is generated appropriately for the PR. If you edited the file noted in the previous step, this should run lint on currency and tests on currency and components. The tests should pass (so long as your edit didn't change anything). There will be a skipped job for handling manual dispatches.
5. Manually dispatch the `ci.yml` workflow using the PR number you created in the previous step. It should be executed in the same manner as it was from the PR, with the exception of an additional step that pulls PR information.
6. Add a repository variable to your fork from `https://github.com/YOURFORK/woocommerce/settings/variables/actions` named `SKIP_CI` with a truthy value (I use "1").
7. Create a new PR similar to step 3. This time, the CI triggered by `ci.yml` should be skipped.
8. Run the CI manually by providing the PR number from step 7. The CI should run when triggered via dispatch.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

**Review by rrennick — DISMISSED** [review]
Nice work on this @jonathansadowski. My test PR was https://github.com/rrennick/woocommerce/pull/6. I closed then opened the PR to re-trigger CI. My manual run logs can be seen at https://github.com/rrennick/woocommerce/actions/workflows/ci.yml.

**Comment by adimoldovan** [comment]
@woocommerce/vortex, I solved the conflicts in this PR and also fixed some integration conflicts. This can be reviewed and tested agin.

Here are my test steps:
1. Create a fork of this repo, and merge the changes in the PR to trunk within your fork.
2. Enable actions if they aren't already. You may want to disable other workflows that run on the PR so that you don't have to wait around for ci.yml to receive workers.
3. Create a new PR against trunk of your fork. Edit a file within one of the projects or plugins covered by ci.yml. For example, make an edit (such as adding a comment) to packages/js/currency/src/index.ts.

Created PR https://github.com/adimoldovan/woocommerce/pull/1

4. Verify that the project matrix is generated appropriately for the PR. If you edited the file noted in the previous step, this should run lint on currency and tests on currency and components. The tests should pass (so long as your edit didn't change anything). There will be a skipped job for handling manual dispatches.

See run https://github.com/adimoldovan/woocommerce/actions/runs/9746767866.
All expected jobs ran.

5. Manually dispatch the ci.yml workflow using the PR number you created in the previous step. It should be executed in the same manner as it was from the PR, with the exception of an additional step that pulls PR information.

See run https://github.com/adimoldovan/woocommerce/actions/runs/9746797439.
All expected jobs ran, same as in step 4.

6. Add a repository variable to your fork from https://github.com/YOURFORK/woocommerce/settings/variables/actions named SKIP_CI with a truthy value (I use "1").
7. Create a new PR similar to step 3. This time, the CI triggered by ci.yml should be skipped.

See run https://github.com/adimoldovan/woocommerce/actions/runs/9747170235.
No jobs ran.
The [Evaluate Project Job Statuses](https://github.com/adimoldovan/woocommerce/actions/runs/9747170235/job/26899223952#logs) fails as expected, because the `project-jobs` job needs to run and be successful.

8. Run the CI manually by providing the PR number from step 7. The CI should run when triggered via dispatch.

See run https://github.com/adimoldovan/woocommerce/actions/runs/9747187636/job/26899288366
All expected jobs ran.

**Comment by adimoldovan** [comment]
Another `ci.yml` conflicts solving followed by another round of testing:

update PR without SKIP_CI: https://github.com/adimoldovan/woocommerce/actions/runs/9776367516
sim PR without SKIP_CI: https://github.com/adimoldovan/woocommerce/actions/runs/9776527620
update PR with SKIP_CI: https://github.com/adimoldovan/woocommerce/actions/runs/9776558592
sim PR with SKIP_CI: https://github.com/adimoldovan/woocommerce/actions/runs/9776698077

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -39,16 +78,24 @@ jobs:
...
         name: 'Build Matrix'
         id: 'project-jobs'
+        env:
+          PR_SIM: ${{ needs.dispatch-handler.outputs.base }}
         with:
           script: |
-            let baseRef = ${{ toJson( github.base_ref ) }};
+            const prSim = process.env.PR_SIM;
+            
+            let baseRef = prSim || ${{ toJson( github.base_ref ) }};
```
Comment (kalessil) [inline .github/workflows/ci.yml:98]: Out of curiosity, why don't we wrap `prSim` into `toJson`?

```
    let baseRef = ${{ toJson( needs.dispatch-handler.outputs.base || github.base_ref ) }};
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -39,16 +78,24 @@ jobs:
         name: 'Checkout'
         with:
           fetch-depth: 0
+          # If the workflow wasn't triggered by dispatch, this will be empty and use defaults.
+          ref: ${{ needs.dispatch-handler.outputs.head }}
```
Comment (kalessil) [inline .github/workflows/ci.yml:82]: Will it properly reference the head in both WooCommerce and fork repositories?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -71,23 +118,35 @@ jobs:
...
       matrix:
         include: ${{ fromJSON( needs.project-jobs.outputs.lint-jobs ) }}
+
     steps:
       - uses: 'actions/checkout@v4'
         name: 'Checkout'
         with:
           fetch-depth: 0
+          # If the workflow wasn't triggered by dispatch, this will be empty and use defaults.
+          ref: ${{ needs.dispatch-handler.outputs.head }}
```
Comment (kalessil) [inline .github/workflows/ci.yml:149]: Will it properly reference the head in both WooCommerce and fork repositories?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -98,20 +157,27 @@ jobs:
...
       fail-fast: false
       matrix:
         include: ${{ fromJSON( needs.project-jobs.outputs.test-jobs ) }}
+
     steps:
       - uses: 'actions/checkout@v4'
         name: 'Checkout'
+        with:
+          # If the workflow wasn't triggered by dispatch, this will be empty and use defaults.
+          ref: ${{ needs.dispatch-handler.outputs.head }}
```
Comment (kalessil) [inline .github/workflows/ci.yml:179]: Will it properly reference the head in both WooCommerce and fork repositories?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (adimoldovan) [reply .github/workflows/ci.yml:98]: I guess that should work as well and looks more elegant. It's Jonathan's initial implementation, so I cannot answer on his thinking. 
I would rather not make such changes at this point, that will require another round of time consuming testing in a fork.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (adimoldovan) [reply .github/workflows/ci.yml:82]: From my testing it looks like it does. Did you find it doesn't?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (kalessil) [reply .github/workflows/ci.yml:98]: Gotcha! If it works, we are ok to keep it as it is.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (kalessil) [reply .github/workflows/ci.yml:82]: Not yet, did first review sweep to understand changes and manual testing TBD yet.

Comment (rrennick) [reply .github/workflows/ci.yml:82]: > Will it properly reference the head in both WooCommerce and fork repositories?

In PRs, `head` is the target branch in the same repository.

**Comment by kalessil** [comment]
@adimoldovan: followed the testing instructions and everything worked except the  `Evaluate Project Job Statuses` job.

Changes in  packages/js/currency/src/index.ts as suggested in testing instructions.

Test PR in a fork: https://github.com/kalessil/woocommerce/pull/1
- Handle dispatched workflow: skipped
- Matrix looks good overall, but one of dependant package tests didn't run (@woocommerce/product-editor) while another did (@woocommerce/components)

Dispatched run: https://github.com/kalessil/woocommerce/actions/runs/9857792022/job/27217641618
- Handle dispatched workflow: executed
- Same set of jobs is executed (@woocommerce/product-editor tests are not running as well)

SKIP_CI: https://github.com/kalessil/woocommerce/pull/2
- the jobs are skipped

Evaluate Project Job Statuses: failed

```
Run # Check if project-jobs was successful. Fail for any other status, including skipped. 
  # Check if project-jobs was successful. Fail for any other status, including skipped. 
  result="skipped"
  if [[ $result != "success" ]]; then
    echo "Generating CI jobs was not successful."
    exit 1
  fi
  
  node .github/workflows/scripts/evaluate-jobs-conclusions.js
  shell: /usr/bin/bash -e {0}
  env:
    FORCE_COLOR: 1
    REPOSITORY: kalessil/woocommerce
    RUN_ID: 9857917825
    GITHUB_TOKEN: ***
Generating CI jobs was not successful.
Error: Process completed with exit code 1.
```

**Comment by adimoldovan** [comment]
> @adimoldovan: followed the testing instructions and everything worked except the `Evaluate Project Job Statuses` job.

Thank you @kalessil. That job is expected to fail in that case, we want to block merges if no checks ran.

**Pull request merged**